src/fstext/determinize-star-inl.h

```diff
@@ -468,99 +469,221 @@ template<class Arc> class DeterminizerStar {
+    // e.g. if the original set is (s1, s2)
+    // s1 -eps-> s4
+    // s2 -eps-> s3 -eps-> s4
+    // then the depth for s4 would eventually be 2
+
+    // to avoid duplicate of computation, we use a priority_queue for storing
+    // the state_id's to process, and always pick the one with the smallest
+    // depth so far first
+    //
+    // this would be similar to topo-sort the states before processing;
+    // (not quite identical since we don't compute the definitive depth first,
+    // but gradually update the depth while running
+    // it seems to speed up the process very significantly already
+    int depth;
+  };
+
+  // this is the struct that is put into the priority_queue. 
+  // The comparison operator makes sure that the queue.top() returns
+  // the one with the least depth
+  struct IdAndDepth {
+    IdAndDepth (InputStateId i, int d):
+      id(i), depth(d) {}
+    InputStateId id;
+    int depth;
+    bool operator <(const IdAndDepth& other) const {
+      return this->depth > other.depth;  // we want the lowest depth on top
+    }
+  };
+  
+  // queue for EpsilonClosure() function
+  // put it here so the memeory allocation for better effiency
+  // otherwise it gets created and destroyed constantly, also
+  // possibly a lot of resize() call the for underlying vector
+  priority_queue<IdAndDepth> queue;
+
+  // Sub-routine that we call in EpsilonClosure()
+  // To avoid the function getting too long, we make it a separate function
+  // It takes the current "unprocessed_weight" and propagate it to the 
+  // states accessible from elem.state by an epsilon arc
+  // and add the results to cur_subset.
+  void ExpandOneElement(const Element &elem,
+                        bool sorted,
+                        Weight unprocessed_weight,
+                        std::map<InputStateId, EpsilonClosureInfo> *cur_subset,
+                        int this_depth) {
+    // now we are going to propagate the "unprocessed_weight"
+    for (ArcIterator<Fst<Arc> > aiter(*ifst_, elem.state);
+         !aiter.Done(); aiter.Next()) {
+      const Arc &arc = aiter.Value();
+      if (sorted && arc.ilabel > 0) {
+        break;
+        // Break from the loop: due to sorting there will be no
+        // more transitions with epsilons as input labels.
+      }
+      if (arc.ilabel == 0) {  // Epsilon transition.
+        Element next_elem;
+        next_elem.state = arc.nextstate;
+        next_elem.weight = Weight::Zero();
+        Weight next_unprocessed_weight
+                       = Times(unprocessed_weight, arc.weight);
+
+        // now must append strings
+        if (arc.olabel == 0)
+          next_elem.string = elem.string;
+        else {
+          vector<Label> seq;
+          repository_.SeqOfId(elem.string, &seq);
+          if (arc.olabel != 0)
+            seq.push_back(arc.olabel);
+          next_elem.string = repository_.IdOfSeq(seq);
+        }
```

Comment:
Hainan - about the problem we were trying to solve with the 'in_queue' variable... I think I may have a way to solve this.
If you take the rest of this function starting from this line (and this function is anyway too long), and make it a function, e.g. AddOneElement(const Element &element, int32 depth);
then you could call that same function when iterating over input_subset [i.e. AddOneElement(input_subset[i], 0)].
Although this would have an O(n log n) lookup, you already have such a thing when creating the initial map, so there is no loss-- this way avoid having to have two such lookups.